executor/src/witgen/machines/block_machine.rs

```diff
@@ -153,13 +153,37 @@ impl<T: FieldElement> Machine<T> for BlockMachine<T> {
     fn witness_col_values(&mut self, fixed_data: &FixedData<T>) -> HashMap<String, Vec<T>> {
         std::mem::take(&mut self.data)
             .into_iter()
-            .map(|(id, values)| {
-                let mut values = values
+            .map(|(id, mut values)| {
+
+                // For all constraints to be satisfied, unused cells have to be filled with valid values.
+                // We do this, we construct a default block, by repeating the first input to the block machine.
+
+                if values.len() < 2 * self.block_size {
+                    log::warn!("Block machine was never used. Assuming that filling witness columns with zeros satisfies the constraints.");
+                }
+
+                values.resize(fixed_data.degree as usize, None);
+
+                let second_block_values = values
+                    .iter()
+                    .skip(self.block_size)
+                    .take(self.block_size);
+
+                // The first block is a dummy block (filled mostly with None), the second block is the first block
+                // resulting of an actual evaluation.
+                // However, if the block machine already sets some registers in the last row of the previous block,
+                // they will be set in the "dummy block". In this case, we want to use these values.
+                // As a result, the default block consists of values of the first block if they are set, otherwise
+                // the values of the second block.
```

Comment:
Please add a TODO that we need to determine the row-extend per column at some point (also for other uses in that file).